### ISSUE TITLE
[Sealed Types] Permits clause handling in grammar/parser is unnecessarily complicated

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#2652]
1. Here are the current grammar rules:

```

Goal ::= RestrictedIdentifierpermits PermittedSubclasses

ClassHeaderPermittedSubclassesopt ::= $empty
ClassHeaderPermittedSubclassesopt -> ClassHeaderPermittedSubclasses
/:$readableName ClassHeaderPermittedSubclasses:/
/:$compliance 15:/

-- Production name hardcoded in parser. Must be ::= and not ->
PermittedSubclasses ::= ClassTypeList
/:$readableName PermittedSubclasses:/

ClassHeaderPermittedSubclasses ::= RestrictedIdentifierpermits ClassTypeList
/.$putCase consumeClassHeaderPermittedSubclasses(); $break ./
/:$readableName ClassHeaderPermittedSubclasses:/
/:$compliance 15:/

InterfaceHeaderPermittedSubClassesAndSubInterfacesopt ::= $empty
InterfaceHeaderPermittedSubClassesAndSubInterfacesopt -> InterfaceHeaderPermittedSubClassesAndSubInterfaces
/:$readableName InterfaceHeaderPermittedSubClassesAndSubInterfaces:/
/:$compliance 15:/

InterfaceHeaderPermittedSubClassesAndSubInterfaces ::= RestrictedIdentifierpermits ClassTypeList
/.$putCase consumeInterfaceHeaderPermittedSubClassesAndSubInterfaces(); $break ./
/:$readableName InterfaceHeaderPermittedSubClassesAndSubInterfaces:/
/:$compliance 15:/
```

This looks needlessly elaborate. (also note compliance should be 17 not 15)

The following grammar should suffice if properly hooked up.

```
PermittedTypesopt -> $empty
PermittedTypesopt ::= 'permits' ClassTypeList
/.$putCase consumePermittedTypes(); $break ./
/:$readableName PermittedTypes:/
/:$compliance 17:/
```

2. On the scanner side too, disambiguating 'permits' the contextual keyword from the identifier 'permits' is overly complicated. Given the very specific context with unambiguous prefix where 'permits' may feature as a contextual keyword, it ought to suffice to ask the parser whether it would shift the keyword 'permits' to disambiguate.

**Comment by srikanth-sankaran** [comment]
> 2. On the scanner side too, disambiguating 'permits' the contextual keyword from the identifier 'permits' is overly complicated. Given the very specific context with unambiguous prefix where 'permits' may feature as a contextual keyword, it ought to suffice to ask the parser whether it would shift the keyword 'permits' to disambiguate.

In https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2672 I am reversing the scanner changes. The scanner simplifications are not incorrect, they simply don't play well when the CU has syntax errors - in terms of recovery and repair.

**Comment by srikanth-sankaran** [comment]
This fix has been reverted for 4.33 M1 - See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2672
